### ISSUE TITLE
SameLen not inferred for CharSequence

### ISSUE THREAD

**Issue by panacekcz** [kelloggm/checker-framework#210]
When allocating an array of a length same as a `CharSequence`, `@SameLen` is not inferred.
Code:
```java
public class CharSequenceSameLen {
	public void m(CharSequence s) {
		char[] c = new char[s.length()];
		for(int i=0;i<s.length();++i) {
			c[i] = s.charAt(i);
		}
	}
}
```

Output:
```
CharSequenceSameLen.java:5: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
			c[i] = s.charAt(i);
			  ^
  found   : @LTLengthOf(value="s", offset="0") int
  required: @IndexFor("c") or @LTLengthOf("c") -- an integer less than c's length
1 error
```

Expected: no warnings.

Similar code with `String` works. The solution is to treat `CharSequence.length()` the same as `String.length()` in the index checker implementation.

Should we do that for all methods overriding `CharSequence.length()`? It would be useful for custom fixed-size collections implementing `CharSequence` . On the other hand, we do not want that for `StringBuilder`.

**Comment by kelloggm** [comment]
In my opinion, we should implement a general method for allowing SameLen inference. Then, we could remove the special handling for Strings and use that method anywhere. My impression is that this is what's requested in #194.

**Comment by panacekcz** [comment]
Yes. On the other hand, fixing this would be easy compared to that.

**Comment by kelloggm** [comment]
This was fixed when #194 was fixed. Closing now, but will commit the testcase above with all the other tests from these issues in a batch.
